plugins/modules/ec2_vpc_dhcp_option.py

```diff
@@ -190,91 +245,233 @@
+            module.fail_json_aws(e, msg="Unable to delete dhcp option {0}".format(dhcp_options_id))
+
+    return changed
 
 
-def fetch_dhcp_options_for_vpc(vpc_conn, vpc_id):
+def match_dhcp_options(client, module, new_config):
     """
-    Returns the DHCP options object currently associated with the requested VPC ID using the VPC
-    connection variable.
+    Returns a DhcpOptionsId if the module parameters match; else None
+    Filter by tags, if any are specified
     """
-    vpcs = vpc_conn.get_all_vpcs(vpc_ids=[vpc_id])
-    if len(vpcs) != 1 or vpcs[0].dhcp_options_id == "default":
-        return None
-    dhcp_options = vpc_conn.get_all_dhcp_options(dhcp_options_ids=[vpcs[0].dhcp_options_id])
-    if len(dhcp_options) != 1:
-        return None
-    return dhcp_options[0]
+    try:
+        all_dhcp_options = client.describe_dhcp_options(aws_retry=True)
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Unable to describe dhcp options")
+
+    for dopts in all_dhcp_options['DhcpOptions']:
+        if module.params['tags']:
+            # If we were given tags, try to match on them
+            boto_tags = ansible_dict_to_boto3_tag_list(module.params['tags'])
+            if dopts['DhcpConfigurations'] == new_config and dopts['Tags'] == boto_tags:
+                return True, dopts['DhcpOptionsId']
+        elif dopts['DhcpConfigurations'] == new_config:
+            return True, dopts['DhcpOptionsId']
+
+    return False, None
 
 
-def match_dhcp_options(vpc_conn, tags=None, options=None):
+def create_dhcp_config(module):
     """
-    Finds a DHCP Options object that optionally matches the tags and options provided
+    Convert provided parameters into a DhcpConfigurations list that conforms to what the API returns:
+    https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeDhcpOptions.html
+        [{'Key': 'domain-name',
+         'Values': [{'Value': 'us-west-2.compute.internal'}]},
+        {'Key': 'domain-name-servers',
+         'Values': [{'Value': 'AmazonProvidedDNS'}]},
+         ...],
     """
-    dhcp_options = vpc_conn.get_all_dhcp_options()
-    for dopts in dhcp_options:
-        if (not tags) or get_resource_tags(vpc_conn, dopts.id) == tags:
-            if (not options) or dopts.options == options:
-                return(True, dopts)
-    return(False, None)
-
-
-def remove_dhcp_options_by_id(vpc_conn, dhcp_options_id):
-    associations = vpc_conn.get_all_vpcs(filters={'dhcpOptionsId': dhcp_options_id})
-    if len(associations) > 0:
-        return False
-    else:
-        vpc_conn.delete_dhcp_options(dhcp_options_id)
-        return True
+    new_config = []
+    params = module.params
+    if params['domain_name'] is not None:
+        new_config.append({'Key': 'domain-name', 'Values': [{'Value': params['domain_name']}]})
+    if params['dns_servers'] is not None:
```

Comment:
what about just `params.get('domain_name'):` and reduce `if` and `for`in single `for server in params.get('dns_servers'):` 
but yeah, then the extra vars (`dns_server_list`, `ntp_server_list` etc are always be initialized, even if not used.

Reply:
Historically the default value of those options is None, not an empty list, and we can't iterate on None.  If I change the default to [] I get cascading effects in the data transformations where null values get set on options that should be unset, which breaks changed status assertions.  It should be possible to make this change but it would require changing code in several places unfortunately.